File: nguilib/src/main/java/com/ng/nguilib/PointLoadingView.java
package com.ng.nguilib;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.AnimatorSet;
import android.animation.ValueAnimator;
import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;
import android.view.animation.AccelerateInterpolator;
import android.view.animation.DecelerateInterpolator;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.ng.nguilib.utils.DensityUtil;
import com.ng.nguilib.utils.LogUtils;

import java.util.ArrayList;
import java.util.List;

/**
 * 描述:
 *
 * @author Jzn
 * @date 2019-12-10
 */
public class PointLoadingView extends View {
    private final int STATE_RUNNING = 0x1;
    private final int STATE_STOP = 0x2;
    private static int mAnimState = 0x2;

    private float mHeight;
    private float[] mPointHeight;
    private float mWidth;
    private float mPointSpace;

    //time
    private static final long TIME_ALL_INTERVAL = 300;
    private static final long TIME_POINT_INTERVAL = 200;
    private static final long TIME_DOWN_ANIM = 300;
    private static final long TIME_UP_ANIM = 600;

    //R of point
    private static int mPointDiameter;

    //default point size
    private static final int POINT_SIZE = 3;

    //tools
    private Paint mPointPaint;
    private Paint mBgPaint;
    private AnimatorSet mDownAnimSet;
    private AnimatorSet mUpAnimSet;

    //color
    private static final int DEFAULT_BG_COLOR = Color.WHITE;
    private static final int DEFAULT_POINT_COLOR = Color.BLACK;

    public void setColor(int bgColor, int pointColor) {
        mBgPaint.setColor(bgColor);
        mPointPaint.setColor(pointColor);

    }

    public PointLoadingView(Context context) {
        super(context);
        init(context);
    }

    public PointLoadingView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }


    private void init(Context context) {
        mAnimState = STATE_STOP;
        mPointDiameter = DensityUtil.INSTANCE.dip2px(context, 6f);
        mPointHeight = new float[POINT_SIZE];
        initPaint();
    }

    private void initPaint() {
        mPointPaint = new Paint();
        mPointPaint.setColor(DEFAULT_POINT_COLOR);
        mPointPaint.setStrokeWidth(mPointDiameter);
        mPointPaint.setAntiAlias(true);
        mPointPaint.setStrokeCap(Paint.Cap.ROUND);
        mPointPaint.setStyle(Paint.Style.FILL);
        mBgPaint = new Paint();
        mBgPaint.setAntiAlias(true);
        mBgPaint.setStyle(Paint.Style.FILL);
        mBgPaint.setColor(DEFAULT_BG_COLOR);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        drawBg(canvas);
        drawPoints(canvas);
    }

    private void drawBg(Canvas canvas) {
        RectF bgRectF = new RectF(0, 0, mWidth, mHeight);
        canvas.drawRoundRect(bgRectF, mHeight / 2, mHeight / 2, mBgPaint);
    }

    private void drawPoints(Canvas canvas) {
        for (int i = 0; i < POINT_SIZE; i++) {
            canvas.drawPoint((i + 1) * mPointSpace, mPointHeight[i], mPointPaint);
        }
    }


    private void initAnim() {
        //down
        if (mDownAnimSet == null) {
            mDownAnimSet = new AnimatorSet();
            List<Animator> mDownAnimList = new ArrayList<>();
            for (int i = 0; i < POINT_SIZE; i++) {
                ValueAnimator tempDownAmt = ValueAnimator.ofFloat(0f, 100f);
                tempDownAmt.setDuration(TIME_DOWN_ANIM);
                tempDownAmt.setInterpolator(new AccelerateInterpolator());
                tempDownAmt.setStartDelay(i * TIME_POINT_INTERVAL);
                final int finalI = i;
                tempDownAmt.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
                    @Override
                    public void onAnimationUpdate(ValueAnimator animation) {
                        float temp = animation.getAnimatedFraction();
                        mPointHeight[finalI] = mHeight / 2 + (mHeight / 2 + mPointDiameter) * temp;
                        postInvalidate();
                    }
                });
                mDownAnimList.add(tempDownAmt);
            }
            mDownAnimSet.playTogether(mDownAnimList);
            mDownAnimSet.setStartDelay(TIME_ALL_INTERVAL);
            mDownAnimSet.addListener(new AnimatorListenerAdapter() {
                @Override
                public void onAnimationEnd(Animator animation) {
                    super.onAnimationEnd(animation);
                    if (mAnimState != STATE_STOP)
                        mUpAnimSet.start();
                }
            });
        }

        //up
        if (mUpAnimSet == null) {
            mUpAnimSet = new AnimatorSet();
            List<Animator> mUpAnimList = new ArrayList<>();
            for (int i = 0; i < POINT_SIZE; i++) {
                ValueAnimator tempUpAmt = ValueAnimator.ofFloat(0f, 100f);
                tempUpAmt.setDuration(TIME_UP_ANIM);
                tempUpAmt.setInterpolator(new DecelerateInterpolator());
                tempUpAmt.setStartDelay(i * TIME_POINT_INTERVAL);
                final int finalI = i;
                tempUpAmt.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
                    @Override
                    public void onAnimationUpdate(ValueAnimator animation) {
                        float temp = animation.getAnimatedFraction();
                        mPointHeight[finalI] = mHeight + mPointDiameter - (mHeight / 2 + mPointDiameter) * temp;
                        postInvalidate();
                    }
                });
                mUpAnimList.add(tempUpAmt);
            }
            mUpAnimSet.playTogether(mUpAnimList);
            mUpAnimSet.setStartDelay(TIME_ALL_INTERVAL);
            mUpAnimSet.addListener(new AnimatorListenerAdapter() {
                @Override
                public void onAnimationEnd(Animator animation) {
                    super.onAnimationEnd(animation);
                    if (mAnimState != STATE_STOP)
                        mDownAnimSet.start();
                }
            });
        }
    }


    public void startLoadingAnim() {
        if (mDownAnimSet != null && mAnimState == STATE_STOP) {
            mAnimState = STATE_RUNNING;
            mDownAnimSet.start();
        }
    }

    public void stopLoadingAnim() {
        mAnimState = STATE_STOP;
        if (mDownAnimSet != null) {
            mDownAnimSet.cancel();
            mDownAnimSet.end();
        }
        if (mUpAnimSet != null) {
            mUpAnimSet.cancel();
            mUpAnimSet.end();
        }
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        stopLoadingAnim();
    }

    @Override
    protected void onVisibilityChanged(@NonNull View changedView, int visibility) {
        super.onVisibilityChanged(changedView, visibility);
        logd("visibility: " + visibility);
        if (visibility == VISIBLE) {
            startLoadingAnim();
        } else {
            stopLoadingAnim();
        }

    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        mHeight = getMeasuredHeight();
        mWidth = getMeasuredWidth();
        mPointSpace = mWidth / (POINT_SIZE + 1);
        for (int i = 0; i < POINT_SIZE; i++) {
            mPointHeight[i] = mHeight / 2;
        }
        initAnim();
        logd(mHeight + "   " + mWidth + "   " + mPointSpace);
    }

    public void logd(String txt) {
        Log.d("nangua", txt);
    }
}

File: nguilib/src/main/java/com/ng/nguilib/view/PointLoadingView.kt
package com.ng.nguilib.view

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.ValueAnimator
import android.view.animation.DecelerateInterpolator
import android.animation.AnimatorSet
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.view.animation.AccelerateInterpolator
import android.graphics.RectF
import android.util.Log
import com.ng.nguilib.utils.DensityUtil


/**
 * 描述:
 * @author Jzn
 * @date 2019-12-14
 */
class PointLoadingView(context: Context, attrs: AttributeSet) : View(context, attrs) {
    private val STATE_RUNNING = 0x1
    private val STATE_STOP = 0x2
    private var mAnimState = 0x2

    private var mHeight: Float = 0.toFloat()
    private var mPointHeight: FloatArray? = null
    private var mWidth: Float = 0.toFloat()
    private var mPointSpace: Float = 0.toFloat()

    //time
    private val TIME_ALL_INTERVAL: Long = 300
    private val TIME_POINT_INTERVAL: Long = 200
    private val TIME_DOWN_ANIM: Long = 300
    private val TIME_UP_ANIM: Long = 600

    //R of point
    private var mPointDiameter: Float = 0f
    private val POINT_DIAMETER = 12f


    //default point size
    private val POINT_SIZE = 3


    //tools
    private var mPointPaint: Paint = Paint()
    private var mBgPaint: Paint = Paint()
    private var mDownAnimSet: AnimatorSet? = null
    private var mUpAnimSet: AnimatorSet? = null

    //color
    private val DEFAULT_BG_COLOR = Color.WHITE
    private val DEFAULT_POINT_COLOR = Color.BLACK


    fun setColor(bgColor: Int, pointColor: Int) {
        mBgPaint.color = bgColor
        mPointPaint.color = pointColor

    }

    init {
        init(getContext())
    }


    private fun init(context: Context) {
        mAnimState = STATE_STOP
        mPointDiameter = DensityUtil.dip2px(context, POINT_DIAMETER).toFloat()
        mPointHeight = FloatArray(POINT_SIZE)
        initPaint()
    }

    private fun initPaint() {
        mPointPaint = Paint()
        mPointPaint.color = DEFAULT_POINT_COLOR
        mPointPaint.strokeWidth = mPointDiameter
        mPointPaint.isAntiAlias = true
        mPointPaint.strokeCap = Paint.Cap.ROUND
        mPointPaint.style = Paint.Style.FILL
        mBgPaint.isAntiAlias = true
        mBgPaint.style = Paint.Style.FILL
        mBgPaint.color = DEFAULT_BG_COLOR
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        drawBg(canvas)
        drawPoints(canvas)
    }

    private fun drawBg(canvas: Canvas) {
        val bgRectF = RectF(0f, 0f, mWidth, mHeight)
        canvas.drawRoundRect(bgRectF, mHeight / 2, mHeight / 2, this.mBgPaint)
    }

    private fun drawPoints(canvas: Canvas) {
        for (i in 0 until POINT_SIZE) {
            canvas.drawPoint((i + 1) * mPointSpace, mPointHeight!![i], mPointPaint)
        }
    }


    private fun initAnim() {
        logd("initAnim")
        //down
        if (mDownAnimSet == null) {
            mDownAnimSet = AnimatorSet()
            val mDownAnimList = ArrayList<ValueAnimator>()
            for (i in 0 until POINT_SIZE) {
                val tempDownAmt = ValueAnimator.ofFloat(0f, 100f)
                tempDownAmt.duration = TIME_DOWN_ANIM
                tempDownAmt.interpolator = AccelerateInterpolator()
                tempDownAmt.startDelay = i * TIME_POINT_INTERVAL
                tempDownAmt.addUpdateListener { animation ->
                    val temp = animation.animatedFraction
                    mPointHeight?.set(i, mHeight / 2 + (mHeight / 2 + mPointDiameter) * temp)
                    postInvalidate()
                }
                mDownAnimList.add(tempDownAmt)
            }
            mDownAnimSet!!.playTogether(mDownAnimList as Collection<Animator>?)
            mDownAnimSet!!.startDelay = TIME_ALL_INTERVAL
            mDownAnimSet!!.addListener(object : AnimatorListenerAdapter() {
                override fun onAnimationEnd(animation: Animator) {
                    super.onAnimationEnd(animation)
                    if (mAnimState != STATE_STOP)
                        mUpAnimSet!!.start()
                }
            })
        }
        //up
        if (mUpAnimSet == null) {
            mUpAnimSet = AnimatorSet()
            val mUpAnimList = ArrayList<ValueAnimator>()
            for (i in 0 until POINT_SIZE) {
                val tempUpAmt = ValueAnimator.ofFloat(0f, 100f)
                tempUpAmt.duration = TIME_UP_ANIM
                tempUpAmt.interpolator = DecelerateInterpolator()
                tempUpAmt.startDelay = i * TIME_POINT_INTERVAL
                tempUpAmt.addUpdateListener { animation ->
                    val temp = animation.animatedFraction
                    mPointHeight?.set(i, mHeight + mPointDiameter - (mHeight / 2 + mPointDiameter) * temp)
                    postInvalidate()
                }
                mUpAnimList.add(tempUpAmt)
            }
            mUpAnimSet!!.playTogether(mUpAnimList as Collection<Animator>?)
            mUpAnimSet!!.startDelay = TIME_ALL_INTERVAL
            mUpAnimSet!!.addListener(object : AnimatorListenerAdapter() {
                override fun onAnimationEnd(animation: Animator) {
                    super.onAnimationEnd(animation)
                    if (mAnimState != STATE_STOP)
                        mDownAnimSet!!.start()
                }
            })
        }
    }


    fun startLoadingAnim() {
        initAnim();
        if (mDownAnimSet != null && mAnimState == STATE_STOP) {
            mAnimState = STATE_RUNNING
            mDownAnimSet!!.start()
        }
    }

    fun stopLoadingAnim() {
        mAnimState = STATE_STOP
        if (mDownAnimSet != null) {
            mDownAnimSet!!.cancel()
            mDownAnimSet!!.end()
        }
        if (mUpAnimSet != null) {
            mUpAnimSet!!.cancel()
            mUpAnimSet!!.end()
        }
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        stopLoadingAnim()
    }

    override fun onVisibilityChanged(changedView: View, visibility: Int) {
        super.onVisibilityChanged(changedView, visibility)
        logd("visibility: $visibility")
        if (visibility == VISIBLE) {
            startLoadingAnim()
        } else {
            stopLoadingAnim()
        }

    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)

    }

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        super.onLayout(changed, left, top, right, bottom)
        mHeight = measuredHeight.toFloat()
        mWidth = measuredWidth.toFloat()
        mPointSpace = mWidth / (POINT_SIZE + 1)
        for (i in 0 until POINT_SIZE) {
            mPointHeight?.set(i, mHeight / 2)
        }
        initAnim()
        logd("$mHeight   $mWidth   $mPointSpace")
    }

    fun logd(txt: String) {
        Log.d("nangua", txt)
    }
}